# Module 2: Data Analytics with Python - Advanced Statistics

## Lab 2.1: Advance Statistics


## Objevtive
***
- **Hypothesis testing**
    - Level of significance
    - One Tail Test
    - Two-tailed test
    - p-value
    

- **Compute the test statistic**

     - z-Test
         - One sampled test
         - Two sampled test
     - t-Test
         - One sampled test
         - Two sampled test
         - Paired test
     - Chi-Square Test
     - ANOVA(Analysis Of Varience Test)




<br><br>
## **z-Test in hypothesis testing**
***

   The z-test is a hypothesis test in which the z-statistic follows a normal distribution. The z-test is best used for greater-than-30 samples because, under the central limit theorem, as the number of samples gets larger, the samples are approximately normally distributed.
    Z test is performed when standard deviation is known.

**One Sample Z test:**
    
   In one sample z-test, we are checking one data group and deciding whether sample mean is equal to population mean or not.
   Blood pressure can be measured with a person's arm or finger. The dataset used here has two columns which are the readings of systolic blood pressure of patients using these two methods. The two columns are named: armsys, fingsys.
   
  In the one sample z-test we are using armsys group and deciding whether the sample mean is same as the population mean or not.


## **Hypothesis Testing**
***
Hypothesis testing or significance testing is a statistical method for testing a claim or hypothesis about a parameter in a population, using data measured in a sample.
Hypothesis Testing is basically an assumption that we make about the population parameter. E.g., Average marks obtained by students is 60.

Steps to follow in Hypothesis Testing:

**1.	State the hypothesis**

We have two parameters here: Null Hypothesis and Alternate Hypothesis.
The Null Hypothesis assumes that there is no relation between two variables i.e., one variable has no effect on another variable. The Null Hypothesis is a basic assumption made based on domain or problem knowledge.
The alternative hypothesis is the hypothesis used in hypothesis testing that is contrary to the null hypothesis.
NOTE: If the null hypothesis is rejected, then we accept the alternative hypothesis. If the null hypothesis is not rejected, then we do not accept the alternative hypothesis.

**2.	Set the criteria for a decision**

To set the criteria for a decision, we state the level of significance for a test.
Confidence level is something that how confident we are regarding the given statement. Generally, we won’t be 100% sure so it will be less than 100%. 

   Level of significance:
        Refers to the degree of significance in which we accept or reject the null-hypothesis. Level of significance is denoted by α.
        α= 1-confidence level

   One Tail Test:

        A test of a statistical hypothesis, where the region of rejection is on only one side of the sampling distribution, is called a one-tailed test.
        Ex:  Average marks of students <= 40

   Two-tailed test:

        A two-tailed test is a statistical test in which the critical area of a distribution is two-sided and tests whether a sample is greater than or less than a certain range of values. If the sample being tested falls into either of the critical areas, the alternative hypothesis is accepted instead of the null hypothesis.
        Example: Average marks of students! = 40 students

   p-value:

        A p value is used in hypothesis testing to help you support or reject the null hypothesis. The p value is the evidence against a null hypothesis. The smaller the p-value, the stronger the evidence that you should reject the null hypothesis.

**NOTE: If your P value is less than the chosen significance level then you reject the null hypothesis i.e., accept that your sample gives reasonable evidence to support the alternative hypothesis.**
<br><br>
**3.	Compute the test statistic**

These are most widely used Tests for Hypothesis Testing.
- Z Test
- T Test
- ANOVA Test
- Chi-Square Test

**4.	Make a decision based on test results.**


In [1]:
import pandas as pd
from scipy import stats

url = "https://raw.githubusercontent.com/bluedataconsulting/AIMasteryProgram/main/Lab_Exercises/Module2/systolic%20blood%20pressure.csv"
df = pd.read_csv(url)



In [2]:
df.tail(5)

armsys  fingsys
195     122      136
196     100      112
197     136      124
198     124      120
199     100      106

In [3]:
print("Mean   : {}".format(df["armsys"].mean()))
print("Median : {}".format(df["armsys"].median()))
print("Min    : {}".format(df["armsys"].min()))
print("Max    : {}".format(df["armsys"].max()))

Mean   : 128.52
Median : 125.0
Min    : 79
Max    : 220


In [4]:
from statsmodels.stats import weightstats as stests

ztest, pval = stests.ztest(df['armsys'], value=125)#value = mean of 'armsys' under null hypothesis
print(pval)

if pval<0.05:                                # 0.05 is the value of α
    print("Mean is not 125")
else:
    print("Mean is 125")

import numpy as np

0.032546442541153964
Mean is not 125


**Conclusion: Hence, we accept the alternative hypothesis.**

<br><br>
**Two Sample z-test**

In two sample z-test, we are checking two independent data groups and deciding whether sample mean of two group is equal or not.

In [5]:
import pandas as pd
from scipy import stats

url = "https://raw.githubusercontent.com/bluedataconsulting/AIMasteryProgram/main/Lab_Exercises/Module2/systolic%20blood%20pressure.csv"
df = pd.read_csv(url)

from statsmodels.stats import weightstats as stests

ztest, pval = stests.ztest(x1=df['armsys'], x2=df['fingsys'], value=0,alternative='two-sided')
print(pval)

if pval<0.05:
  print("Mean of two samples is not 0")
else:
  print("Mean of two samples is 0")

0.07954652069053099
Mean of two samples is 0


**Conclusion: Hence, we accept the null hypothesis.**

<br><br>
## **t-Test in hypothesis testing**
***

A t-Test is a type of inferential statistic which is used to determine if there is a significant difference between the means of two groups which may be related in certain features.
The t-Test is used to compare the mean of two given groups. The sample follows the Gaussian distribution. A t-Test is used when parameters like the standard deviation of the population are not known.



 **One sample t-Test**
 
The mass of a sample of n=20 are m = 8.8, 6.6, 9.5, 11.2, 10.2, 7.4, 8.0, 9.6, 9.9, 9.0, 7.6, 7.4, 10.4, 11.1, 8.5, 10.0, 11.6, 10.7, 10.3, and 7.0 g.We need to check if there is any difference between the average mass of this sample and the average mass of all acorns of μ = 10.0 g.

In [6]:
#One Sample t test
x = [8.8, 6.6, 9.5, 11.2, 10.2, 7.4, 8.0, 9.6, 9.9, 9.0,
     7.6, 7.4, 10.4, 11.1, 8.5, 10.0, 11.6, 10.7, 10.3, 7.0]
mu = 10
t_critical = 2.093
x_bar = np.array(x).mean()
s = np.array(x).std(ddof=1) # subtract 1 from N to get unbiased estimate of sample standard deviation
N = len(x)
SE = s/np.sqrt(N)
t = (x_bar - mu)/SE
print("t-statistic: ",t)
# A one sample t-test that gives you the p-value too can be done with scipy as follows:
t, p = stats.ttest_1samp(x, mu)
print("t = ", t, ", p = ", p)
if p<0.05:
    print("Reject null hypothesis")
else:
    print("Accept null hypothesis")


t-statistic:  -2.2491611580763977
t =  -2.2491611580763973 , p =  0.03655562279112415
Reject null hypothesis


**Conclusion: We reject the null hypothesis.**

<br><br>
### Two sample t-test
***
 
Null Hypothesis: There is no difference between the two-sample means

Alternate Hypothesis: There is a difference between the two-sample means

Suppose we take N1 = 20 samples and N2 = 30 samples of the weight of acorns


In [7]:
x1 = [10.8, 10.0, 8.2, 9.9, 11.6, 10.1, 11.3, 10.3, 10.7, 9.7, 
      7.8, 9.6, 9.7, 11.6, 10.3, 9.8, 12.3, 11.0, 10.4, 10.4]

x2 = [7.8, 7.5, 9.5, 11.7, 8.1, 8.8, 8.8, 7.7, 9.7, 7.0, 
      9.0, 9.7, 11.3, 8.7, 8.8, 10.9, 10.3, 9.6, 8.4, 6.6,
      7.2, 7.6, 11.5, 6.6, 8.6, 10.5, 8.4, 8.5, 10.2, 9.2]

t_critical = 1.677    #given

X1 = len(x1)
X2 = len(x2)

t1 = X1-1
t2 = X2-1

df = t1+t2

s1 = np.std(x1,ddof=1)
s2 = np.std(x2,ddof=1)

x1_bar = np.mean(x1)
x2_bar = np.mean(x2)
sp = np.sqrt((t1*s1**2 + t2*s2**2)/df)
se = sp*np.sqrt(1/X1 + 1/X2)
t = (x2_bar - x1_bar)/(se)
print("t-statistic", t)

# NOTE: the p-value given is two-sided so the one-sided p value would be p/2
t, p_twosided = stats.ttest_ind(x2, x1, equal_var=True)
print("t = ",t, ", p_twosided = ", p_twosided, ", p_onesided =", p_twosided/2)

if p<0.05:
    print("Reject null hypothesis")
else:
    print("Accept null hypothesis")



t-statistic -3.5981947686898033
t =  -3.5981947686898033 , p_twosided =  0.0007560337478801464 , p_onesided = 0.0003780168739400732
Reject null hypothesis



**Conclusion: We reject the null hypothesis.**
<br><br>

### Paired t-test
***

The paired sample t-test is also called a dependent sample t-test. Let’s take an example from a blood pressure dataset. We need to check the sample means of blood pressure of an individual before and after treatment.

Null hypothesis: The mean difference between the two samples is 0.

Alternate hypothesis: The mean difference between the two samples is not 0.


In [8]:
import pandas as pd
from scipy import stats

x1 = [10.8, 10.0, 8.2, 9.9, 11.6, 10.1, 11.3, 10.3, 10.7, 9.7, 
      7.8, 9.6, 9.7, 11.6, 10.3, 9.8, 12.3, 11.0, 10.4, 10.4]

x2 = [7.8, 7.5, 9.5, 11.7, 8.1, 8.8, 8.8, 7.7, 9.7, 7.0, 
      9.0, 9.7, 11.3, 8.7, 8.8, 10.9, 10.3, 9.6, 8.4, 6.6]

ttest,pval = stats.ttest_rel(x1,x2)

if pval<0.05:
    print("reject null hypothesis")
else:
    print("accept null hypothesis")


reject null hypothesis


**Conclusion: Here, we reject Null Hypothesis.**


### **Chi-Square Test**
***

A chi-square statistic is one way to show a relationship between twocategoricalvariables.The distribution of the chi-square statistic is called thechi-square distribution. Thechi-squaredistributionis defined by the following probability density function:

This tests whether two categorical variables are related or independent.

Here we are taking a 2x3 table for chi-square test.




In [9]:
from scipy.stats import chi2_contingency
table = [[10, 20, 30],[6,  9,  17]]
stat, p, dof, expected = chi2_contingency(table)
print('P value:', p)
if p > 0.05:
    print('Independent')
else:
    print('Dependent')

P value: 0.873028283380073
Independent


**Conclusion: As p>0.05, the variables are independent of each other.**

### **ANOVA (Analysis Of Variance) F-test**
***

This test works well when dealing with two groups, but sometimes we want to compare more than two groups at the same time. 
The Analysis of Variance (ANOVA) is a statistical inference test that lets you compare multiple groups at the same time.
F = Variability between groups or Variability within groups

Unlike the z and t-distributions, the F-distribution does not have any negative values because between and within-group variability are always positive due to squaring each deviation.

**One Way F-test (Anova)**

It tells whether two or more groups are similar or not based on their mean similarity and f-score.

p> 0.05 implies they are similar

p< 0.05 implies they are not similar


In [10]:
%matplotlib inline
import pandas as pd
import statsmodels.api as sm
from statsmodels.formula.api import ols
url = "https://raw.githubusercontent.com/bluedataconsulting/AIMasteryProgram/main/Lab_Exercises/Module2/PlantGrowth.csv"
data = pd.read_csv(url)
mod = ols('weight ~ group', data=data).fit()      #we use ols to set up model
aov_table = sm.stats.anova_lm(mod,type=2)

print(aov_table)

            df    sum_sq   mean_sq         F   PR(>F)
group      2.0   3.76634  1.883170  4.846088  0.01591
Residual  27.0  10.49209  0.388596       NaN      NaN


**Conclusion: From AOV_Table we observe that value of p=0.01591 > 0.05 and therefore both groups are similar.**

### Thank You !!!
